### PR TITLE
Log if a server certificate lacks the subjectAlternativeName extensions

### PR DESCRIPTION
**Pull request by amcasey**
# Log if a server certificate lacks the subjectAlternativeName extensions

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

## Description

CommonName was [deprecated](https://developer.chrome.com/blog/chrome-58-deprecations/#remove-support-for-commonname-matching-in-certificates) in favor of subjectAlternativeName so there's a good chance of getting a browser security warning if it's missing.

### CODE REVIEW

**Comment by amcasey** [comment]
Suggested by @JamesNK at the Tuesday meeting.

File: src/Servers/Kestrel/Core/src/Internal/Infrastructure/TransportManager.cs [hunk 1]
```diff
@@ -76,7 +76,7 @@ public async Task<EndPoint> BindAsync(EndPoint endPoint, MultiplexedConnectionDe
         // The QUIC transport will check if TlsConnectionCallbackOptions is missing.
         if (listenOptions.HttpsOptions != null)
         {
-            var sslServerAuthenticationOptions = HttpsConnectionMiddleware.CreateHttp3Options(listenOptions.HttpsOptions);
+            var sslServerAuthenticationOptions = HttpsConnectionMiddleware.CreateHttp3Options(listenOptions.HttpsOptions, Trace);
```
Comment (amcasey) [inline src/Servers/Kestrel/Core/src/Internal/Infrastructure/TransportManager.cs:79]: This is the reason we use `ILogger`, rather than `ILogger<HttpsConnectionMiddleware>`.  Is there a better way?

File: src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs [hunk 2]
```diff
@@ -596,6 +600,9 @@ public static void FoundCertWithPrivateKey(this ILogger<HttpsConnectionMiddlewar
     [LoggerMessage(8, LogLevel.Debug, "Failed to open certificate store {StoreName}.", EventName = "FailToOpenStore")]
     public static partial void FailedToOpenStore(this ILogger<HttpsConnectionMiddleware> logger, string? storeName, Exception exception);
 
+    [LoggerMessage(9, LogLevel.Information, "Certificate with thumbprint {Thumbprint} lacks the subjectAlternativeName (SAN) extension and may not be accepted by browsers.", EventName = "NoSubjectAlternativeName")]
```
Comment (amcasey) [inline src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:603]: How unique do IDs have to be?  I would have guessed per-logger-type, but the non-generic `ILogger` extension methods seem to reuse them freely.

File: src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs [hunk 3]
```diff
@@ -596,6 +600,9 @@ public static void FoundCertWithPrivateKey(this ILogger<HttpsConnectionMiddlewar
...
+    public static partial void NoSubjectAlternativeName(this ILogger logger, string thumbprint);
```
Comment (amcasey) [inline src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:604]: This should really be `ILogger<HttpsConnectionMiddleware>`.

File: src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs (hunk 2 above)
Comment (amcasey) [inline src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:603]: I went with `Information`, rather than `Warning`, because this server might not receive auth requests from browsers (i.e. it might not actually matter that the extension is missing.).

**Review by amcasey — COMMENTED** [review]
Some notes on the parts I don't like.  Suggestions?

File: src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs (hunk 2 above)
Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:603]: Uniqueness is the `ILogger` name +  logger message ID (which is ID and string)

Usually, the logger used by a type has the same name as the type, e.g. `Microsoft.AspNetCore.Kestrel.HttpConnectionMiddleware`. Then all the logs defined here can have their own IDs that start from 1.

tldr as long as they are unique in this file you should be fine.

File: src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs (hunk 3 above)
Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:604]: It doesn't matter other than consistency.

File: src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs (hunk 2 above)
Comment (BrennanConroy) [reply src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:603]: But in this case since you're passing in KestrelTrace you're breaking that assumption. Should probably try to get an `ILogger<HttpsConnectionMiddleware>` in `BindAsync` and pass that in.

Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:603]: Yes. The KestrelTrace logger won't have the name `Microsoft.AspNetCore.Kestrel.HttpConnectionMiddleware`. It has its own `ILogger` name and will already have a log message with ID=9. So that won't be unique.

Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:603]: Info is fine. I think we need to provide more detail about why browsers might not accept it. Is there a good doc or article that we could link to?

> Certificate with thumbprint {Thumbprint} lacks the subjectAlternativeName (SAN) extension. Some browsers require a SAN and will not accept this certificate. For more information, see https://aka.ms/aspnet/cert-san.

Comment (amcasey) [reply src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:603]: I linked an article in the description, but we don't own it.

Comment (amcasey) [reply src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:603]: I found this, but it seems to be about running a CA (and seems nervous about using SANs at all?).
https://learn.microsoft.com/en-us/previous-versions/windows/it-pro/windows-server-2008-R2-and-2008/ff625722(v=ws.10)?redirectedfrom=MSDN#BKMK_Security

Comment (amcasey) [reply src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:603]: @vcsjones Is this (i.e. specifying a subject alternative name) even a sensible thing to warn about / recommend?  Maybe we should only do it for self-signed certs or when the common name is `localhost`?

File: src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs [hunk 4]
```diff
@@ -452,12 +452,16 @@ private static async ValueTask<SslServerAuthenticationOptions> ServerOptionsCall
...
     }
 
-    internal static void EnsureCertificateIsAllowedForServerAuth(X509Certificate2 certificate)
+    internal static void EnsureCertificateIsAllowedForServerAuth(X509Certificate2 certificate, ILogger logger)
     {
         if (!CertificateLoader.IsCertificateAllowedForServerAuth(certificate))
         {
             throw new InvalidOperationException(CoreStrings.FormatInvalidServerCertificateEku(certificate.Thumbprint));
         }
+        else if (!CertificateLoader.DoesCertificateHaveASubjectAlternativeName(certificate))
```
Comment (blowdart) [inline src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:461]: You could go further here. In modern browsers the SAN is selected by default if present and the subject is ignored. If you have a SAN that doesn't contain the subject that may also be log worthy.

File: src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs (hunk 2 above)
Comment (BrennanConroy) [reply src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:603]: > What would be the consequences of sharing a logger across endpoints?

Shouldn't be any, they are already effectively Singletons in DI.

I vote for whatever option gives us an `ILogger<HttpsConnectionMiddleware>`.

Comment (amcasey) [reply src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:603]: Thanks!

Comment (amcasey) [reply src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:603]: @BrennanConroy Tell me more about the expectation that they'll be singletons?  That does seem to be the case for the [default impl](https://source.dot.net/#Microsoft.Extensions.Logging/LoggerFactory.cs,143), but it's not clear from the name ("factory", rather than "provider") or the [doc comment](https://source.dot.net/#Microsoft.Extensions.Logging.Abstractions/ILoggerFactory.cs,15)
```csharp
        /// <summary>
        /// Creates a new <see cref="ILogger"/> instance.
        /// </summary>
        /// <param name="categoryName">The category name for messages produced by the logger.</param>
        /// <returns>The <see cref="ILogger"/>.</returns>
        ILogger CreateLogger(string categoryName);
```

Comment (BrennanConroy) [reply src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:603]: https://source.dot.net/#Microsoft.Extensions.Logging/LoggingServiceCollectionExtensions.cs,38
and
https://source.dot.net/#Microsoft.Extensions.Logging/LoggingServiceCollectionExtensions.cs,39
is where logging gets added for use in ASP.NET Core apps. And AddLogging is called from the host builders.

If you use `ILoggerFactory` it has a cache of loggers that it creates/gets from:
https://source.dot.net/#Microsoft.Extensions.Logging/LoggerFactory.cs,143 and the `ILogger` impl we use, actually just calls the `ILoggerFactory` impl: https://source.dot.net/#Microsoft.Extensions.Logging.Abstractions/LoggerT.cs,26

So at the end of the day it's all singleton loggers.

Comment (blowdart) [reply src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:603]: Even our internal cert issues implemented SANs 10 years+ ago (Because muggins here wrote the request UI).

But whilst the cab forum rules are meant to be "Thou shalt not" there are moments when CAs forget themselves, due to bugs, so a information message is still a good idea.

File: src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs (hunk 4 above)
Comment (amcasey) [reply src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:461]: I think the suggestion here was to log in cases where the SAN doesn't incorporate the CommonName.  Assuming I understood that correctly, are there helpers for checking that?  There seem to be quite a few different ways to specify SANs, including wildcards, and I'm not sure I'd want to write a validator from scratch.  But maybe there's just a helper for "does this name match that SAN"?

File: src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs (hunk 3 above)
Comment (JamesNK) [inline src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:604]: Exact logger type now works. Be consistent with other log methods:
```suggestion
    public static partial void NoSubjectAlternativeName(this ILogger<HttpsConnectionMiddleware> logger, string thumbprint);
```

File: src/Servers/Kestrel/Core/src/CertificateLoader.cs [hunk 5]
```diff
@@ -106,6 +106,9 @@ internal static bool IsCertificateAllowedForServerAuth(X509Certificate2 certific
     internal static bool DoesCertificateHaveAnAccessiblePrivateKey(X509Certificate2 certificate)
         => certificate.HasPrivateKey;
 
+    internal static bool DoesCertificateHaveASubjectAlternativeName(X509Certificate2 certificate)
+        => certificate.Extensions.OfType<X509SubjectAlternativeNameExtension>().Any();
```
[the remark this answers is not in the document]
Comment (amcasey) [reply src/Servers/Kestrel/Core/src/CertificateLoader.cs:110]: I'm also not used to using `OfType` - I copied it from `IsCertificateAllowedForServerAuth`.

I'm probably missing something, but it looks like [the indexer](https://source.dot.net/#System.Security.Cryptography/System/Security/Cryptography/X509Certificates/X509ExtensionCollection.cs,56) does the same linear search?  And it doesn't appear to use the struct enumerator pattern, so writing the foreach loop myself probably wouldn't be a win either?

File: src/Servers/Kestrel/Core/src/KestrelServer.cs [hunk 6]
```diff
@@ -35,7 +35,7 @@ public KestrelServer(IOptions<KestrelServerOptions> options, IConnectionListener
             options,
             new[] { transportFactory ?? throw new ArgumentNullException(nameof(transportFactory)) },
             Array.Empty<IMultiplexedConnectionListenerFactory>(),
-            new SimpleHttpsConfigurationService(),
+            new SimpleHttpsConfigurationService(loggerFactory.CreateLogger<HttpsConnectionMiddleware>()),
```
[the remark this answers is not in the document]
Comment (amcasey) [reply src/Servers/Kestrel/Core/src/KestrelServer.cs:38]: Does that have an advantage beyond being idiomatic?

File: src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs (hunk 3 above)
Comment (amcasey) [reply src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:604]: Just a bad copy-paste on my part.

File: src/Servers/Kestrel/Core/src/KestrelServer.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (amcasey) [reply src/Servers/Kestrel/Core/src/KestrelServer.cs]: Done but I'm still curious about the reasoning.

File: src/Servers/Kestrel/Core/src/CertificateLoader.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/CertificateLoader.cs:110]: The foreach in the indexer is over a `List<X509Extension>`, so will use its struct enumerator.

The indexer looks like it allocates an Oid. There are allocations no matter what we do.

I guess just leave it as is.

**Review by JamesNK — APPROVED** [review]

File: src/Servers/Kestrel/Core/src/KestrelServer.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/KestrelServer.cs]: Usually, either types create their own logger with `ILoggerFactory`, or DI does it for them. Creating a type explicitly and making the creator create the logger feels off.

I don't think there is explicit reason except maybe it's about ownership. `SimpleHttpsConfigurationService` is using the logger so it should create it. IDK 🤷‍♂️

File: src/Servers/Kestrel/Core/src/Internal/Infrastructure/TransportManager.cs (hunk 1 above)
Comment (amcasey) [reply src/Servers/Kestrel/Core/src/Internal/Infrastructure/TransportManager.cs]: No longer using `ILogger`.

File: src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs (hunk 2 above)
Comment (amcasey) [reply src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:603]: I don't suppose either @vcsjones or @blowdart is aware of a page I could link to?  (There are some useful pages online, but I understand aka.ms links must point to 1P properties.)

**Comment by amcasey** [comment]
@BrennanConroy Any remaining concerns?

**Review by BrennanConroy — APPROVED** [review]

File: src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs (hunk 4 above)
Comment (blowdart) [reply src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:461]: I've not see anything, and it's fair if you want to avoid it because it may lead to false positives.

File: src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs (hunk 2 above)
Comment (blowdart) [reply src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs:603]: I don't know of anything 1st party
